File: passenger_seating.py
modul = 1000000007

fin = open ('input.txt', 'r')            
fout = open ('output.txt', 'w')

n, k = map(int, fin.readline().split())

support = [[0] * (n + 1) for i in range(3)]
support[1][0] = 1

pattern = [0 for i in range(n + 1)]

for i in range(k):
    place = int(fin.readline())
    pattern[place] = 1

print(pattern)
for i in range(1, n + 1):
    support[0][i] = (support[1][i - 1] + support[2][i - 1]) % modul
    if pattern[i] != 1:
        support[1][i] = support[0][i - 1] % modul
        support[2][i] = support[1][i - 1] % modul

fout.write(str((support[0][-1] + support[1][-1]) % modul))

File: dp/passenger_seating.cpp
#include <vector>
#include <fstream>
#include <iostream>

#define modul 1000000007

using namespace std;

int main() {
    ifstream fin("input.txt");
    ofstream fout("output.txt");

    int n, k;
    fin >> n >> k;

    int *pattern = new int[n + 1]();

    int place;
    for (int i = 0; i < k; i++) {
        fin >> place;
        pattern[place] = 1;
    }

    int support[n + 1][3];
    for (int i = 0; i < n + 1; i++) {
        for (int j = 0; j < 3; j++) {
            support[i][j] = 0;
        }
    }
    support[0][1] = 1;

    for (int i = 1; i < n + 1; i++) {
        support[i][0] = (support[i - 1][1] + support[i - 1][2]) % modul;
        if (pattern[i] != 1) {
            support[i][1] = support[i - 1][0] % modul;
            support[i][2] = support[i - 1][1] % modul;
        }
    }

    fout << (support[n][0] + support[n][1]) % modul;
}
